// electron-oauth-twitter/lib/OauthTwitter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

var _electron = require('electron');

var _oauth = require('oauth');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _class = function () {
  function _class(_ref) {
    var key = _ref.key,
        secret = _ref.secret;

    _classCallCheck(this, _class);

    (0, _assert2.default)(key, 'OAuth Consumer Key is needed!');
    (0, _assert2.default)(secret, 'OAuth Consumer secret is needed!');
    this.consumerKey = key;
    this.consumerSecret = secret;
    this.window = null;
    this.resolve = null;
    this.reject = null;
  }

  _createClass(_class, [{
    key: 'startRequest',
    value: function startRequest() {
      var _this = this;

      var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      var forceLogin = options.force_login || false;
      var authUrl = 'https://api.twitter.com/oauth/authenticate?force_login=' + forceLogin.toString() + ';oauth_token=';
      var oauth = new _oauth.OAuth('https://api.twitter.com/oauth/request_token', 'https://api.twitter.com/oauth/access_token', this.consumerKey, this.consumerSecret, '1.0A', null, 'HMAC-SHA1');

      var deferredPromise = new Promise(function (resolve, reject) {
        var isResolved = false;
        _this.resolve = function (value) {
          if (isResolved) {
            return;
          }

          isResolved = true;
          resolve(value);
        };

        _this.reject = function (error) {
          if (isResolved) {
            return;
          }

          isResolved = true;
          reject(error);
        };
      });

      oauth.getOAuthRequestToken(function (error, oauthToken, oauthTokenSecret) {
        if (error) {
          _this.reject(error);
          return;
        }

        var url = authUrl + oauthToken;
        _this.getAccessToken(oauth, oauthToken, oauthTokenSecret, url);
      });
      return deferredPromise;
    }
  }, {
    key: 'getAccessToken',
    value: function getAccessToken(oauth, oauthToken, oauthTokenSecret, authUrl) {
      var _this2 = this;

      this.window = new _electron.BrowserWindow({ width: 800, height: 600 });
      this.window.loadURL(authUrl);
      this.window.on('close', function () {
        _this2.reject(new Error('the window is closed before complete the authentication.'));
      });
      this.window.webContents.on('will-navigate', function (event, url) {
        /**
         * If 2fa is set, the url includes challenge_id, challenge_type
         */
        if (url.indexOf('challenge_type') >= 0 && url.indexOf('challenge_id') >= 0) {
          _this2.window.loadURL(url);
        } else {
          var matched = url.match(/\?oauth_token=([^&]*)&oauth_verifier=([^&]*)/);
          if (matched) {
            oauth.getOAuthAccessToken(oauthToken, oauthTokenSecret, matched[2], function (error, oauthAccessTokenSecret, oauthAccessTokenSecretSecret) {
              if (error) {
                _this2.reject(error);
                return;
              }

              _this2.resolve({
                oauth_access_token: oauthAccessTokenSecret,
                oauth_access_token_secret: oauthAccessTokenSecretSecret
              });
              _this2.window.close();
            });
          }
        }
      });
    }
  }]);

  return _class;
}();

exports.default = _class;
module.exports = exports['default'];
